### ISSUE TITLE
Drop 2.6 and 3.3 support

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#1245]
Should we drop support for Python 2.6 or 3.3?

Historically we have targetted these versions because we intend dask to be an infrastructural library and because these versions are fairly easy to support in the core of the library.

We might consider dropping support for these versions in some submodules (like dask.dataframe) where users are typically more up to date than in other modules (like dask.array) or the core.

**Comment by eriknw** [comment]
+1 for at least dropping Python 2.6.

**Comment by jcrist** [comment]
Scikit-image just dropped 2.6 support (https://github.com/scikit-image/scikit-image/pull/2033), and xarray only supports 2.7 and 3.3+, so I think we're fine dropping 2.6 for at least `dask.array`. I don't think we can drop 3.3 yet, as both packages still support 3.3. So I'm also +1 on dropping 2.6.

**Comment by seibert** [comment]
We dropped both Python 2.6 and 3.3 from Numba.  Our conda package download numbers indicated that 3.3 was the least popular, but Python 2.6 was the most annoying to support, so we deprecated both, and then dropped them the following release.

**Comment by bryevdv** [comment]
FWIW Bokeh is also only tested on py 2.7 and 3.4+

**Comment by seibert** [comment]
I should note that we never found a situation where Python 3.3 and 3.4 behaved differently in a way we had to work around, so support for Python 3.3 likely costs you very little effort.  We just wanted to tidy up our 3D configuration matrix.

**Comment by ianozsvald** [comment]
I surveyed our meetup audience (200 folk) at PyDataLondon last night (and I do this most months), we're about 50/50 for Py 2.7 vs 3.4+ in the audience with a slight favour still to using Py2.7. Very few folk in the room use Py2.6 and one or two might be stuck on 2.4 (e.g. old banking systems). For Py 3 most people are on 3.4 or 3.5. Dask is cutting edge, you've got a lot of latitude I'd argue to drop older versions that cause you a drag.

**Comment by mrocklin** [comment]
Thanks for the poll.  That's quite helpful.

**Comment by ianozsvald** [comment]
I'll also note that I ran a similar poll at PyDataBerlin conf 2 weeks back, it was more like a 40/60 split for Python 2/3 (and the majority of each were on 2.7 or 3.4+). I've also been gathering evidence for the dropping of 2.6 support and the switch to 3.4+ only:
http://ianozsvald.com/2016/02/29/will-we-see-module-on-python-3-4-is-free-but-only-paid-support-for-python-2-7/

**Comment by seibert** [comment]
Just to be clear, dropping Python 2.6 is a *very* different question than dropping Python 2.7.  Unless a project absolutely depends on a Python 3 feature (or is purely a hobby, so do what makes you happy), I think Python 2.7 support is still a really good idea, and not onerous with the help of CI.

**Comment by ianozsvald** [comment]
Hello again. I was curious about the question for Python 2vs3 popularity for data scientists. I queried my PyDataLondon audience (3,400+ members) last week, by sending 4 emails I got 13% of them to respond to my survey. Here are the results:
http://ianozsvald.com/2016/06/20/results-for-which-version-of-python-2vs3-do-london-data-scientists-use/

In short - Python 2.7 dominates in business, Python 3.4 dominates outside of work. I hypothesise that we'll hit 50% Python 3.x usage by this time next year for PyDataLondon users. Almost nobody uses Python <=2.6 or Python <=3.3.
